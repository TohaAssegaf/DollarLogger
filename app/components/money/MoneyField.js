import styles from './styles'
import { formatAmbiguousMoneyInput, formatMoneyCents } from '/app/lib/MoneyFormatter'
import parseAmbiguousMoney from '/app/lib/parseAmbiguousMoney'
import React from 'react'
import { Text, TextInput, View } from 'react-native'

export default class MoneyField extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      total: 0
    }
  }

  updateInput(input: string) {
    const formattedInput: string = formatAmbiguousMoneyInput(input)
    if (formattedInput !== input) {
      this.inputField.setNativeProps({ text: formattedInput })
    }
    const total = parseAmbiguousMoney(formattedInput)
    this.setState({ total })
    this.props.onChange(total)
  }

  formatFinalMoneyInput() {
    this.inputField.setNativeProps({ text: formatMoneyCents(this.state.total) })
  }

  render() {
    return (
      <View style={styles.moneyField}>
        <Text style={[styles.currencyCode, this.props.currencyCodeStyles]}>$</Text>
        <TextInput
          style={[styles.textInput, this.props.textInputStyles]}
          keyboardType={"numeric"}
          onChangeText={(input) => this.updateInput(input)}
          value={this.state.input}
          onBlur={() => this.formatFinalMoneyInput()}
          ref={(inputField) => { this.inputField = inputField }}
        />
      </View>
    );
  }
}
