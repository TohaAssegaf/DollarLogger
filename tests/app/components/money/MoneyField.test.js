import MoneyField from '~/app/components/money/MoneyField';
import enzyme from 'enzyme'
import { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import renderer from 'react-test-renderer'

enzyme.configure({ adapter: new Adapter() })
const onChange = jest.fn()

it('renders correctly', () => {
  const rendered = renderer.create(<MoneyField onChange={onChange} />).toJSON();
  expect(rendered).toMatchSnapshot();
});

it('allows valid input', () => {
  const wrapper = shallow(<MoneyField onChange={onChange} />)
  const render = wrapper.dive()
  render.find('TextInput').simulate('changeText', '100')
  expect(wrapper.state().total).toEqual(10000)
  expect(onChange).toHaveBeenLastCalledWith(10000)
});

it('allows ambiguous input', () => {
  const wrapper = shallow(<MoneyField onChange={onChange} />)
  const render = wrapper.dive()
  render.find('TextInput').simulate('changeText', '100.')
  expect(wrapper.state().total).toEqual(10000)
  expect(onChange).toHaveBeenLastCalledWith(10000)
});

it('allows ambiguous . input', () => {
  const wrapper = shallow(<MoneyField onChange={onChange} />)
  const render = wrapper.dive()
  render.find('TextInput').simulate('changeText', '1') // Change to non-zero input first
  render.find('TextInput').simulate('changeText', '.')
  expect(wrapper.state().total).toEqual(0)
  expect(onChange).toHaveBeenLastCalledWith(0)
});

it('allows ambiguous <1 input', () => {
  const wrapper = shallow(<MoneyField onChange={onChange} />)
  const render = wrapper.dive()
  render.find('TextInput').simulate('changeText', '.37')
  expect(wrapper.state().total).toEqual(37)
  expect(onChange).toHaveBeenLastCalledWith(37)
});
