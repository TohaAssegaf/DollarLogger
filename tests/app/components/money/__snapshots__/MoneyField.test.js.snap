// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 20,
        },
        undefined,
      ]
    }
  >
    $
  </Text>
  <TextInput
    allowFontScaling={true}
    keyboardType="numeric"
    onBlur={[Function]}
    onChangeText={[Function]}
    style={
      Array [
        Object {
          "fontSize": 20,
          "minWidth": 100,
        },
        undefined,
      ]
    }
    underlineColorAndroid="rgba(0,0,0,0)"
    value=""
  />
</View>
`;
